firefox_puppeteer/ui/windows.py

```diff
@@ -0,0 +1,479 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this file,
+# You can obtain one at http://mozilla.org/MPL/2.0/.
+
+from time import sleep
+
+from marionette import By, Wait
+from marionette.keys import Keys
+
+import firefox_puppeteer.errors as errors
+
+from ..api.l10n import L10n
+from ..base import BaseLib
+from ..decorators import use_class_as_property
+
+
+class Windows(BaseLib):
+
+    @property
+    def all(self):
+        """Retrieves a list of all open chrome windows.
+
+        :returns: List of :class:`BaseWindow`'s corresponding to the
+                  windows in `marionette.chrome_window_handles`.
+        """
+        return [self.create_window_instance(handle) for handle in
+                self.marionette.chrome_window_handles]
+
+    @property
+    def current(self):
+        """Retrieves the currently selected chrome window.
+
+        :returns: The :class:`BaseWindow` for the currently active window.
+        """
+        return self.create_window_instance(self.marionette.current_chrome_window_handle)
+
+    @property
+    def focused_chrome_window_handle(self):
+        """Returns the currently focused chrome window handle
+
+        :returns: The `window handle` of the focused chrome window.
+        """
+        with self.marionette.using_context('chrome'):
+            return self.marionette.execute_script("""
+              Cu.import("resource://gre/modules/Services.jsm");
+              var win = Services.wm.getMostRecentWindow("");
+              return win.QueryInterface(Ci.nsIInterfaceRequestor)
+                        .getInterface(Ci.nsIDOMWindowUtils)
+                        .outerWindowID.toString();
+            """)
```

Comment:
"handles" from the server are intended to be opaque, but you're using this to compare against a handle from the server, and it works because this is how it's implemented. My conclusion from a recent issue with the server was that for window handles corresponding to tabs we need to re-implement then not to use outerWindowID, but a different ID we generate ourselves. If the same ends up being true for chrome windows, this will break.